### ISSUE TITLE
Java Client: Can't process synchronous request call inside CompletableFuture of another async call 

### ISSUE THREAD

**Issue by ievgen-kolomiiets** [nats-io/nats.java#292]
Version: `io.nats:jnats:2.6.6`
- [x] Defect

#### Steps or code to reproduce the issue:
There are two clients of two different services that use Nats for communication. But one client uses asynchronous API and the second one uses synchronous. 
When the synchronous is called inside of CompletableFuture of asynchronous client it fails to process incoming response.

Here is a Java code that demonstrates the issue:
```java
import io.nats.client.*;

import java.time.*;
import java.util.concurrent.*;

public class Main {
    public static void main(String[] args) throws Exception {
        var conn = Nats.connect("nats://localhost:4222");

        var asyncHelloClient = new AsyncHelloClient(conn);
        var syncWorldClient = new SyncWorldClient(conn);

        var result = asyncHelloClient.sayHello()
            .thenApply(hello -> hello + " " + syncWorldClient.sayWorld())
            .get();

        // Result: "Hello null"
        // Expecting: "Hello World"
        System.out.println(result);

        conn.close();
    }

    private static final class AsyncHelloClient {
        private final Connection connection;
        private final String subject = "Hello";

        private AsyncHelloClient(Connection connection) {
            this.connection = connection;

            connection.createDispatcher(msg -> connection.request(msg.getReplyTo(), subject.getBytes()))
                .subscribe(subject);
        }

        CompletableFuture<String> sayHello() {
            return connection.request(subject, null)
                .thenApply(message -> new String(message.getData()));
        }
    }

    private static final class SyncWorldClient {
        private final Connection connection;
        private final String subject = "World";

        private SyncWorldClient(Connection connection) {
            this.connection = connection;

            connection.createDispatcher(msg -> connection.request(msg.getReplyTo(), subject.getBytes()))
                .subscribe(subject);
        }

        String sayWorld() {
            try {
                var message = connection.request(subject, null, Duration.ofSeconds(5L));
                return message == null ? null : new String(message.getData());
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
    }
}

``` 
#### Expected result:
 Both async and sync clients make successful requests and get the results. 

#### Actual result:
 Sync client gets `null` message after timeout.

**Comment by wallyqs** [comment]
Thanks @ievgen-kolomiiets transferred the issue to nats.java too so could reopen this one too

**Comment by ievgen-kolomiiets** [comment]
Reopened in nats.java

**Comment by kozlovic** [comment]
Sorry for the delay.. I don't understand this code enough to be of a good help, but clearly it seems that the way it is written it "blocks" trying to get the reply for the second request. I can trace and see that "World" is getting back to the connection but can't be dispatched properly. Here is the thread-dump when it fails:
```
"nats:5" #26 prio=5 os_prio=31 cpu=0.32ms elapsed=2.11s tid=0x00007fda40940800 nid=0x9503 waiting on condition  [0x000070000c86f000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@12.0.1/Native Method)
	- parking to wait for  <0x000000043f4009e8> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.parkNanos(java.base@12.0.1/LockSupport.java:235)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@12.0.1/CompletableFuture.java:1869)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@12.0.1/ForkJoinPool.java:3128)
	at java.util.concurrent.CompletableFuture.timedGet(java.base@12.0.1/CompletableFuture.java:1939)
	at java.util.concurrent.CompletableFuture.get(java.base@12.0.1/CompletableFuture.java:2092)
	at io.nats.client.impl.NatsConnection.request(NatsConnection.java:978)
	at io.nats.examples.NatsPub$SyncWorldClient.sayWorld(NatsPub.java:80)
	at io.nats.examples.NatsPub.lambda$main$0(NatsPub.java:40)
	at io.nats.examples.NatsPub$$Lambda$17/0x000000080116d840.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(java.base@12.0.1/CompletableFuture.java:642)
	at java.util.concurrent.CompletableFuture.postComplete(java.base@12.0.1/CompletableFuture.java:506)
	at java.util.concurrent.CompletableFuture.complete(java.base@12.0.1/CompletableFuture.java:2144)
	at io.nats.client.impl.NatsConnection.deliverReply(NatsConnection.java:1051)
	at io.nats.client.impl.NatsConnection.lambda$request$6(NatsConnection.java:1009)
	at io.nats.client.impl.NatsConnection$$Lambda$15/0x000000080116cc40.onMessage(Unknown Source)
	at io.nats.client.impl.NatsDispatcher.run(NatsDispatcher.java:98)
	at java.util.concurrent.Executors$RunnableAdapter.call(java.base@12.0.1/Executors.java:515)
	at java.util.concurrent.FutureTask.run(java.base@12.0.1/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@12.0.1/ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@12.0.1/ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(java.base@12.0.1/Thread.java:835)
```

**Comment by jstapels** [comment]
I ran into this issue as well and was puzzled to figure out what I did wrong. I haven't found the actual cause but I suspect it has to do with calling the CompletableFuture#complete method from within the handling of a response.

A possible solution may be to have the complete method NatsConnection.java:1051 executed within a new task on the executor.
